sql/hive/v4/HiveParser.g4

```diff
@@ -0,0 +1,3075 @@
+    : precedenceUnaryPrefixExpression (precedenceBitwiseXorOperator precedenceUnaryPrefixExpression)*
+    ;
+
+precedenceStarOperator
+    : STAR
+    | DIVIDE
+    | MOD
+    | DIV
+    ;
+
+precedenceStarExpression
+    : precedenceBitwiseXorExpression (precedenceStarOperator precedenceBitwiseXorExpression)*
+    ;
+
+precedencePlusOperator
+    : PLUS
+    | MINUS
+    ;
+
+precedencePlusExpression
+    : precedenceStarExpression (precedencePlusOperator precedenceStarExpression)*
+    ;
+
+precedenceConcatenateOperator
+    : CONCATENATE
+    ;
+
+precedenceConcatenateExpression
+    : precedencePlusExpression (precedenceConcatenateOperator plus=precedencePlusExpression)*
+
+    ;
+
+precedenceAmpersandOperator
+    : AMPERSAND
+    ;
+
+precedenceAmpersandExpression
+    : precedenceConcatenateExpression (precedenceAmpersandOperator precedenceConcatenateExpression)*
+    ;
+
+precedenceBitwiseOrOperator
+    : BITWISEOR
+    ;
+
+precedenceBitwiseOrExpression
+    : precedenceAmpersandExpression (precedenceBitwiseOrOperator precedenceAmpersandExpression)*
+    ;
+
+precedenceRegexpOperator
+    : KW_LIKE
+    | KW_RLIKE
+    | KW_REGEXP
+    ;
+
+precedenceSimilarOperator
+    : precedenceRegexpOperator
+    | LESSTHANOREQUALTO
+    | LESSTHAN
+    | GREATERTHANOREQUALTO
+    | GREATERTHAN
+    ;
+
+subQueryExpression
+    : LPAREN selectStatement RPAREN
+    ;
+
+precedenceSimilarExpression
+    : precedenceSimilarExpressionMain
+    | KW_EXISTS subQueryExpression
+    ;
+
+precedenceSimilarExpressionMain
+    : a=precedenceBitwiseOrExpression part=precedenceSimilarExpressionPart?
+    ;
+
+precedenceSimilarExpressionPart
+    : precedenceSimilarOperator equalExpr=precedenceBitwiseOrExpression
+    | precedenceSimilarExpressionAtom
+    | (KW_NOT precedenceSimilarExpressionPartNot)
+    ;
+
+precedenceSimilarExpressionAtom
+    : KW_IN precedenceSimilarExpressionIn
+    | KW_BETWEEN min=precedenceBitwiseOrExpression KW_AND max=precedenceBitwiseOrExpression
+    | KW_LIKE KW_ANY expr=expressionsInParenthesis
+    | KW_LIKE KW_ALL expr=expressionsInParenthesis
+    | precedenceSimilarExpressionQuantifierPredicate
+    ;
+
+precedenceSimilarExpressionQuantifierPredicate
+    : subQuerySelectorOperator quantifierType subQueryExpression
+    ;
+
+quantifierType
+    : KW_ANY
+    | KW_SOME
+    | KW_ALL
+    ;
+
+precedenceSimilarExpressionIn
+    : subQueryExpression
+    | expr=expressionsInParenthesis
+    ;
+
+precedenceSimilarExpressionPartNot
+    : precedenceRegexpOperator notExpr=precedenceBitwiseOrExpression
+    | precedenceSimilarExpressionAtom
+    ;
+
+precedenceDistinctOperator
+    : KW_IS KW_DISTINCT KW_FROM
+    ;
+
+precedenceEqualOperator
+    : EQUAL
+    | EQUAL_NS
+    | NOTEQUAL
+    | KW_IS KW_NOT KW_DISTINCT KW_FROM
+    ;
+
+precedenceEqualExpression
+    : precedenceSimilarExpression
+        (equal=precedenceEqualOperator p=precedenceSimilarExpression | dist=precedenceDistinctOperator p=precedenceSimilarExpression)*
```

Comment:
Probably it makes sense to use `+=` operator for labels because expression is under closure.

```suggestion
        (equal+=precedenceEqualOperator p+=precedenceSimilarExpression | dist+=precedenceDistinctOperator p+=precedenceSimilarExpression)*
```